p4-fusion/thread_pool.cc

```diff
@@ -170,6 +161,113 @@ ThreadPool::ThreadPool(const int size, const std::string& repoPath, const int tz
 	}
 }
 
+void ThreadPool::startSignalHandlingThread()
+{
+	auto start = [this]()
+	{
+		// Block signals from being handled by the main thread, and all future threads.
+		//
+		// - SIGINT, SIGTERM, SIGHUP: thread pool will be shutdown and std::exit will be called
+		// - SIGUSR1: only sent by ~SignalHandler() to tell the signal handler thread to exit
+		sigset_t blockedSignals;
+		sigemptyset(&blockedSignals);
+		for (auto sig : { SIGINT, SIGTERM, SIGHUP, SIGUSR1 })
+		{
+			sigaddset(&blockedSignals, sig);
+		}
+
+		int rc = pthread_sigmask(SIG_BLOCK, &blockedSignals, nullptr);
+		if (rc != 0)
+		{
+			std::ostringstream oss;
+			oss << "(signal handler) failed to block signals: (" << errno << ") " << strerror(errno);
+			throw std::runtime_error(oss.str());
+		}
+
+		sigset_t signalsToWaitOn = blockedSignals;
+		// Spawn a thread to handle signals.
+		// The thread will block and wait for signals to arrive and then shutdown the thread pool, unless it receives SIGUSR1,
+		// in which case it will just exit (since main() is handling the shutdown).
+		//
+		// Using a separate thread for purely signal handling allows us to use non-reentrant functions
+		// (such as std::cout, condition variables, etc.) in the signal handler.
+		signalHandlingThread = ThreadRAII(std::thread(
+		    [signalsToWaitOn, this]()
+		    {
+			    // Wait for signals to arrive.
+			    int sig;
+			    int rc = sigwait(&signalsToWaitOn, &sig);
+			    if (rc != 0)
+			    {
+				    ERR("(signal handler) failed to wait for signals: (" << errno << ") " << strerror(errno))
+				    ShutDown();
+				    std::exit(errno);
+			    }
+
+			    // Did main() tell us to shutdown?
+			    if (sig == SIGUSR1)
+			    {
+				    // Yes, so no need to print anything - just exit.
+				    return;
+			    }
+
+			    // Otherwise, we received a signal from the OS - print a message and shutdown.
+			    if (!sigismember(&signalsToWaitOn, sig))
+			    {
+				    ERR("(signal handler): WARNING: received signal (" << sig << ") \"" << strsignal(sig) << "\" that is not blocked, this should not happen and indicates a logic error in the signal handler.")
+			    }
+
+			    ERR("(signal handler) received signal (" << sig << ") \"" << strsignal(sig) << "\", shutting down")
+			    ShutDown();
+			    std::exit(sig);
+		    }));
+	};
+
+	std::call_once(m_startSignalHandlingThread_Flag, start);
+}
+
+void ThreadPool::shutdownSignalHandlingThread()
+{
+	auto stop = [this]()
+	{
+		auto errcode = pthread_kill(signalHandlingThread.get().native_handle(), SIGUSR1);
+		if (errcode != 0)
+		{
+			ERR("(signal handler) failed to shut down signal handling thread: (" << errcode << ") " << strerror(errcode))
+			return;
+		}
+
+		SUCCESS("Signal handler shut down successfully")
+		return;
+	};
+
+	return std::call_once(m_shutdownSignalHandlingThread_Flag, stop);
+}
```

Comment:
in the previous implementation we didn't need to signal to the thread because by calling the CV and clearing out the exceptions, it would just return and the destructor of ThreadRAII would join on it then - did you intentionally make this more explicit here?